README.md
What's this?
==================================
vagrant snapshot management plugin
can handle multi-vm environment.

## Install

    gem install vagrant-snap

## Following commands are added

    Tasks:
      vagrant snap back                   # back to current snapshot
      vagrant snap delete SNAP_NAME       # delete snapshot
      vagrant snap go SNAP_NAME           # go to specified snapshot
      vagrant snap help [COMMAND]         # Describe subcommands or one specific subcommand
      vagrant snap list                   # list snapshot
      vagrant snap take [NAME] [-d DESC]  # take snapshot

## Screen capture

![vagrant_snap_list](https://github.com/t9md/vagrant-snap/raw/master/misc/vagrant_snap_list.png)

## Example

In actual example, current snapshot is highlighted with yellow ANSI color.

    t510 dev/vagtest02% vagrant snap take
    [db]
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    [web]
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    t510 dev/vagtest02% 
    t510 [1] dev/vagtest02% vagrant snap list    
    [db]
    +-db-01 [ 16 seconds ]
    [web]
    +-web-01 [ 14 seconds ]
    t510 dev/vagtest02% vagrant snap take -d "2nd snap"
    [db]
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    [web]
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    t510 dev/vagtest02% vagrant snap list 
    [db]
    +-db-01 [ 48 seconds ]
        +-db-02 [ 10 seconds ] 2nd snap
    [web]
    +-web-01 [ 47 seconds ]
        +-web-02 [ 8 seconds ] 2nd snap
    t510 dev/vagtest02% vagrant snap go web-01 web                 
    [web]
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    Restoring snapshot 283d90aa-ef75-4316-a847-e04961c2ec26
    0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    Waiting for the VM to power on...
    VM has been successfully started.
    t510 dev/vagtest02% vagrant snap list 
    [db]
    +-db-01 [ 1 minute ]
        +-db-02 [ 40 seconds ] 2nd snap
    [web]
    +-web-01 [ 1 minute ]
        +-web-02 [ 38 seconds ] 2nd snap
    t510 dev/vagtest02% 


## Similar project
[sahara](https://github.com/jedi4ever/sahara)

## Other
I intentionally avoided naming this plugin as 'vagrant-snapshot', because I believe Vagrant author  
implement snap shot management feature and want to use 'snapshot' as command name.

## VM configuration
In my experience, to avoid `VERR_SSM_LOAD_CONFIG_MISMATCH` error when restoreing snaphot,  
disable `USB controller` and `absolte pointing device`.


CHANGELOG
* 0.10
2012-01-26
   * rewite to support vagrant 0.9(it no longer use Thor).
   * dirty hack but work..
* 0.09
2011-07-19
   * don't die on VM defined in env but not yet created, thanks to Ches Martin
* 0.08
2011-07-18
   * No change but only increment version number for mistake in release.
* 0.07
2011-07-18
   * snapshot name is now start from 0 instead of 001 for less key stroke.
2011-07-18 Fix following bug, thanks to Ches Martin
   * Trying to create a snap with a name given when no previous snapshots exist failed.
   * Spaces in the name of a VM or in a snapshot name passed as -n 'initial build' caused failures.


lib/vagrant_snap.rb
require "virtualbox"
require "colored"
require "pp"

module Snap
  module VBox
    class SnapShot #{{{
      class << self
        def tree
          @@tree
        end

        Snap = Struct.new(:name, :time_stamp, :description, :uuid, :current)

        def init
          @@current = nil
          @@tree = nil
        end

        def parse_tree(vmname)
          init
          vm = ::VirtualBox::VM.find( vmname )
          @@current = vm.current_snapshot
          return unless @@current
          @@tree = _parse(vm.root_snapshot)
        end

        # [TODO] need refactoring
        def time_elapse(time)
          _sec  = 1
          _min  = _sec * 60
          _hour = _min * 60
          _day  = _hour * 24

          sec = time.to_i
          min = sec / _min
          hour = sec / _hour
          day  = sec / _day

          case
          when day  > 0 then "#{day} day#{day == 1 ? '' : 's'}"
          when hour > 0 then "#{hour} hour#{hour == 1 ? '' : 's'}"
          when min  > 0 then "#{min} minute#{min == 1 ? '' : 's'}"
          when sec  > 0 then "#{sec} second#{sec == 1 ? '' : 's'}"
          end
        end

        def _parse(s)
          tree = [ Snap.new(s.name , s.time_stamp, s.description, s.uuid, s.uuid == @@current.uuid) ]
          s.children.each do |c|
            tree.concat [_parse(c)]
          end
          tree
        end

        def format(guide, s)
          time     = time_elapse(Time.now - s.time_stamp)
          snapinfo = "#{s.name} [ #{time} ]"
          snapinfo  = snapinfo.yellow  if s.current
          result   = "#{guide} #{snapinfo}"
          result  << " #{s.description}" unless s.description.empty?
          result  << "\n"
        end

        def lastname
          if tree
            tree.flatten.sort_by(&:time_stamp).last.name
          else
            nil
          end
        end

        def next_available_snapname
          if lastname.nil?
            "0"
          else
            n = lastname.succ
            n = n.succ while VBox::SnapShot.include? n
            n
          end
        end

        def include?(name)
          return false unless tree
          tree.flatten.map(&:name).include? name
        end

        def show(t=tree, guide="")
          result = ""
          t.each_with_index do |v, idx|
            case v
            when Array
              tmp = guide.dup.chop.chop.sub("`", " ") << "    "
              tmp << "#{t.size == idx + 1 ? '`' : '|'}" << "--"
              result << show(v, tmp)
            else
              result << format(guide, v)
            end
          end
          result
        end
      end
    end #}}}
  end

  class Snap < Vagrant::Command::Base
    def help
      puts <<-EOS
Usage: vagrant snap <subcommand> [options...]

 subcommands are..
    vagrant snap list
    vagrant snap back
    vagrant snap go <snapshot> [boxname]
    vagrant snap take [TARGET] [-n SNAP_NAME] [-d DESC]"
    vagrant snap delete <snapshot> [boxname]"
    
      EOS
    end
    def execute# {{{
      @main_args, @sub_command, @sub_args = split_main_and_subcommand(@argv)
      # p [@main_args, @sub_command, @sub_args]

      # sub_list = %w(list go back take delete test)
      sub_list = %w(list go back take delete)
      unless @sub_command
        help
        exit
      end
      if sub_list.include? @sub_command
        send(@sub_command)
      else
        ui.warn "unkown command '#{@sub_command}'"
        help
        exit
      end
    end# }}}

    private
    # def env
    # @_env ||= Vagrant::Environment.new
    # end

    def ui# {{{
      @ui ||= ::Vagrant::UI::Colored.new("vagrant")
    end# }}}
    def safe_with_target_vms(target, &blk)# {{{
      with_target_vms(target) do |vm|
        unless vm.created?
          @logger.info("not created yet: #{vm.name}")
          next
        end
        puts "[#{vm.name}]"
        blk.call(vm)
      end
    end# }}}
    def target_vmname# {{{
      target = @sub_args.empty? ? nil : @sub_args.last.to_s
    end# }}}
    def exe(cmd)# {{{
      puts "# exe: #{cmd}" if @exe_verbose
      system cmd
    end# }}}

    def list# {{{
      # options = {}
      # opts = OptionParser.new { |opts| opts.banner = "vagrant snap list" }
      # argv = parse_options(opts)
      # # p argv
      safe_with_target_vms(target_vmname) do |vm|
        VBox::SnapShot.parse_tree( vm.uuid )
        puts VBox::SnapShot.tree ? VBox::SnapShot.show : "no snapshot"
      end
    end# }}}
    def go # {{{
      options = {}
      opts = OptionParser.new do |opts|
        opts.banner = "Usage: vagrant snapt go <snapshot> [boxname]"
      end
      snapshot, target = *@sub_args
      unless snapshot
        puts opts.help
        return
      end
      safe_with_target_vms(target) do |vm|
        VBox::SnapShot.parse_tree( vm.uuid )
        if VBox::SnapShot.include?( snapshot )
          exe "VBoxManage controlvm '#{vm.uuid}' poweroff"
          exe "VBoxManage snapshot  '#{vm.uuid}' restore '#{snapshot}'"
          exe "VBoxManage startvm   '#{vm.uuid}' --type headless"
        else
          ui.warn "'#{snapshot}' is not exist"
        end
      end
    end# }}}
    def back# {{{
      safe_with_target_vms(target_vmname) do |vm|
        exe "VBoxManage controlvm '#{vm.uuid}' poweroff"
        exe "VBoxManage snapshot  '#{vm.uuid}' restorecurrent"
        exe "VBoxManage startvm   '#{vm.uuid}' --type headless"
      end
    end# }}}
    def take# {{{
      options = {}
      opts = OptionParser.new do |opts|
        opts.banner = "Usage: vagrant snap take [TARGET] [-n SNAP_NAME] [-d DESC]"
        opts.on("-n", "--name STR", "Name of snapshot"       ){ |v| options[:name] = v }
        opts.on("-d", "--desc STR", "Description of snapshot"){ |v| options[:desc] = v }
      end

      begin
        argv =  parse_options(opts)
      rescue OptionParser::MissingArgument
        raise ::Vagrant::Errors::CLIInvalidOptions, :help => opts.help.chomp
      end
      return if !argv
      @main_args, @sub_command, @sub_args = split_main_and_subcommand(argv)
      # p @sub_args
      # return
      # snapshot, target = *@sub_args
      safe_with_target_vms(target_vmname) do |vm|
        VBox::SnapShot.parse_tree( vm.uuid )
        if options[:name] and VBox::SnapShot.include? options[:name]
          ui.warn "'#{options[:name]}' is already exist"
          next
        end
        snapshot = options[:name] ? options[:name] : VBox::SnapShot.next_available_snapname
        cmd = "VBoxManage snapshot '#{vm.uuid}' take '#{snapshot}' --pause"
        if options[:desc]
          cmd << " --description '#{options[:desc]}'"
        end
        exe cmd
      end
    end# }}}

    def delete# {{{
      options = {}
      opts = OptionParser.new do |opts|
        opts.banner = "vagrant snap delete <snapshot> [boxname]"
      end
      snapshot, target = *@sub_args
      unless snapshot
        puts opts.help
        return
      end
      snapshot, target = *@sub_args
      safe_with_target_vms(target) do |vm|
        VBox::SnapShot.parse_tree( vm.uuid )
        if VBox::SnapShot.include?( snapshot )
          exe "VBoxManage snapshot '#{vm.uuid}' delete '#{snapshot}'"
        else
          ui.warn "'#{snapshot}' is not exist"
        end
      end
    end# }}}
  end
end
Vagrant.commands.register(:snap) { ::Snap::Snap }


lib/vagrant_snap.rb_.rb
require "colored"

module Snap
  module VBox
    class SnapShot #{{{
      class << self
        def tree
          @@tree
        end

        Snap = Struct.new(:name, :time_stamp, :description, :uuid, :current)

        def init
          @@current = nil
          @@tree = nil
        end

        def parse_tree(vmname)
          init
          vm = VirtualBox::VM.find( vmname )
          @@current = vm.current_snapshot
          return unless @@current
          @@tree = _parse(vm.root_snapshot)
        end

        # [TODO] need refactoring
        def time_elapse(time)
          _sec  = 1
          _min  = _sec * 60
          _hour = _min * 60
          _day  = _hour * 24

          sec = time.to_i
          min = sec / _min
          hour = sec / _hour
          day  = sec / _day

          case
          when day  > 0 then "#{day} day#{day == 1 ? '' : 's'}"
          when hour > 0 then "#{hour} hour#{hour == 1 ? '' : 's'}"
          when min  > 0 then "#{min} minute#{min == 1 ? '' : 's'}"
          when sec  > 0 then "#{sec} second#{sec == 1 ? '' : 's'}"
          end
        end

        def _parse(s)
          tree = [ Snap.new(s.name , s.time_stamp, s.description, s.uuid, s.uuid == @@current.uuid) ]
          s.children.each do |c|
            tree.concat [_parse(c)]
          end
          tree
        end

        def format(guide, s)
          time     = time_elapse(Time.now - s.time_stamp)
          snapinfo = "#{s.name} [ #{time} ]"
          snapinfo  = snapinfo.yellow  if s.current
          result   = "#{guide} #{snapinfo}"
          result  << " #{s.description}" unless s.description.empty?
          result  << "\n"
        end

        def lastname
          if tree
            tree.flatten.sort_by(&:time_stamp).last.name
          else
            nil
          end
        end

        def include?(name)
          return false unless tree
          tree.flatten.map(&:name).include? name
        end

        def show(t=tree, guide="")
          result = ""
          t.each_with_index do |v, idx|
            case v
            when Array
              tmp = guide.dup.chop.chop.sub("`", " ") << "    "
              tmp << "#{t.size == idx + 1 ? '`' : '|'}" << "--"
              result << show(v, tmp)
            else
              result << format(guide, v)
            end
          end
          result
        end
      end
    end #}}}
  end

  class Command < Vagrant::Command::GroupBase
    register "snap","Manages a snap"

    no_tasks {
      def env
        @_env ||= Vagrant::Environment.new
      end

      def with_target(target, &blk)
        target_found = false
        env.vms.each do |name, vm|
          next if vm.vm.nil?    # not yet created
          vagvmname = vm.name
          vmname    = vm.vm.name

          if target.nil? or target.to_sym == vagvmname
            blk.call(vmname, vagvmname)
            target_found = true
          end
        end
        warn "A VM by the name of `#{target}' was not found".red unless target_found
      end
    }

    desc "list", "list snapshot"
    def list(target=nil)
      with_target(target) do |vmname, vagvmname|
        puts "[#{vagvmname}]"
        VBox::SnapShot.parse_tree( vmname )
        if VBox::SnapShot.tree
          result = VBox::SnapShot.show
        else
          result = "no snapshot"
        end
        puts result
      end
    end

    desc "go SNAP_NAME", "go to specified snapshot"
    def go(snapshot_name, target=nil)
      with_target(target) do |vmname, vagvmname|
        puts "[#{vagvmname}]"
        VBox::SnapShot.parse_tree( vmname )
        if VBox::SnapShot.include?( snapshot_name )
          system "VBoxManage controlvm #{vmname} poweroff"
          system "VBoxManage snapshot  #{vmname} restore #{snapshot_name}"
          system "VBoxManage startvm   #{vmname} --type headless"
        else
          warn "#{snapshot_name} is not exist".red
        end
      end
    end

    desc "back", "back to current snapshot"
    def back(target=nil)
      with_target(target) do |vmname, vagvmname|
        puts "[#{vagvmname}]"
        system "VBoxManage controlvm #{vmname} poweroff"
        system "VBoxManage snapshot  #{vmname} restorecurrent"
        system "VBoxManage startvm   #{vmname} --type headless"
      end
    end

    desc "take [TARGET] [-n SNAP_NAME] [-d DESC]", "take snapshot"
    method_option :desc, :type => :string, :aliases => "-d"
    method_option :name, :type => :string, :aliases => "-n"
    def take(target=nil)
      with_target(target) do |vmname, vagvmname|
        puts "[#{vagvmname}]"
        VBox::SnapShot.parse_tree( vmname )
        if options.name
          if VBox::SnapShot.include? options.name
            warn "#{options.name} is already exist".red
            next
          else
            new_name = options.name
          end
        end
        unless new_name
          lastname = VBox::SnapShot.lastname
          new_name = if lastname.nil?
                       "0"
                     else
                       n = lastname.succ
                       n = n.succ while VBox::SnapShot.include? n
                       n
                     end
        end
        desc = options.desc ? " --description '#{options.desc}'" : ""
        system "VBoxManage snapshot '#{vmname}' take '#{new_name}' #{desc} --pause"
      end
    end

    desc "delete SNAP_NAME", "delete snapshot"
    def delete(snapshot_name, target=nil)
      with_target(target) do |vmname, vagvmname|
        puts "[#{vagvmname}]"
        VBox::SnapShot.parse_tree( vmname )
        if VBox::SnapShot.include?( snapshot_name )
          system "VBoxManage snapshot #{vmname} delete #{snapshot_name}"
        else
          warn "#{snapshot_name} is not exist".red
        end
      end
    end
  end
end
